### ISSUE TITLE
all: ensure that Go toolchain meets Apple’s notarization requirements

### ISSUE THREAD

**Issue by artyom** [golang/go#34986]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.13.3 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/artyom/Library/Caches/go-build"
GOENV="/Users/artyom/Library/Application Support/go/env"
GOEXE=""
GOFLAGS="-ldflags=-w -trimpath"
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/artyom/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/Users/artyom/Library/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/artyom/Library/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/artyom/Repositories/artyom/rex/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/lb/3rk8rqs53czgb4v35w_342xc0000gn/T/go-build499526677=/tmp/go-build -gno-record-gcc-switches -fno-common"

</pre></details>

### What did you do?

Installed go from https://dl.google.com/go/go1.13.3.darwin-amd64.tar.gz and then tried to use it to build project. This is on macOS Catalina 10.15:

    Darwin MacBook-Air.local 19.0.0 Darwin Kernel Version 19.0.0: Wed Sep 25 20:18:50 PDT 2019; root:xnu-6153.11.26~2/RELEASE_X86_64 x86_64

### What did you expect to see?

Go works as usual.

### What did you see instead?

On first `go` call I'm presented with OS pop-up that "macOS cannot verify that this app is free from malware", with options either to discard this window (the process is killed then) or move the binary to trash. After whitelisting this binary via Preferences → Security & Privacy → General → Allow Anyway, `go` can be called fine.

Next I tried to build project with `go install` and got the came pop-up for "compile" tool.

<img width="668" alt="Screen Shot 2019-10-18 at 13 00 17" src="https://user-images.githubusercontent.com/68182/67085837-41aa7280-f1a8-11e9-8e4f-3b8e84a69a85.png">

<img width="668" alt="Screen Shot 2019-10-18 at 13 00 23" src="https://user-images.githubusercontent.com/68182/67085835-41aa7280-f1a8-11e9-9388-8fdbcf447cb1.png">

**Comment by artyom** [comment]
To be able to run updated Go 1.13.3 and compile my project with it, I had to whitelist so far these binaries:

* go
* compile
* asm
* cgo
* link

Whitelisting each takes the following:

1. presented with os pop-up about binary not being verified;
2. click cancel, at this moment binary is killed;
3. go to  Preferences → Security & Privacy → General, click "Allow Anyway" button;
4. try running the same command again;
5. presented with os pop-up about binary not being verified, but now with another option to run binary anyway;
6. click "open" button on that pop-up.

**Comment by bradfitz** [comment]
Good times.

/cc @golang/osp-team

**Comment by bradfitz** [comment]
Related: #34748 (we need a Catalina builder)

**Comment by networkimprov** [comment]
Discussion on Hacker News: https://news.ycombinator.com/item?id=21179970

Also, Catalina is reported to be a disastrous upgrade.

**Comment by andybons** [comment]
I’ll take a look at this.

**Comment by dmitshur** [comment]
I can reproduce on my personal MBP with macOS 10.15 (19A602) when installing from the .pkg installer too, however it's a different (less intrusive) experience.

The "macOS cannot verify that this app is free from malware" prompt appears when first running the installer. After accepting and installing, Go binaries such as `go`, `compile`, do not produce the same prompt. Also, re-running the .pkg installer doesn't ask again.

**Comment by aclements** [comment]
@andybons, I don't think we need any of the runtime exceptions from that page. The only one that we may even potentially want is "Disable Library Validation Entitlement", which would let users use the the `plugin` package without signing their plugins, but it seems like developers probably should sign their plugins.

**Comment by vp2177** [comment]
The (`.pkg`) installer doesn't even start for me. Is that the same issue, or unrelated


![Screenshot 2019-10-18 at 17 10 09](https://user-images.githubusercontent.com/451903/67273609-b76b5280-f4b6-11e9-9871-1f7cc6c0ecfc.png)

**Comment by andybons** [comment]
@vp2177 it’s the same issue.

**Comment by andybons** [comment]
Good news: [according to Apple](https://developer.apple.com/news/?id=09032019a), we don’t have to enable the hardened runtime until January 2020. That said, we’ll likely have to start signing the binaries themselves (not just the installer as we do now), so this issue should stay open even if we fix the issue in the short term via notarization of the installer.

**Comment by andybons** [comment]
@dmitshur to follow up on your experience, do you have SIP (System Integrity Protection) turned on? `csrutil status` should give you that info.

@artyom do you have SIP turned on? Thanks.

**Comment by artyom** [comment]
@andybons yes, it's enabled:

    ¶ csrutil status
    System Integrity Protection status: enabled.

**Comment by andybons** [comment]
Our installer is getting stuck in limbo with the Notarization service. We’ve contacted Apple and are awaiting a response. It’s unclear what the ETA on that will be, though.

**Comment by andybons** [comment]
https://developer.apple.com/system-status/ now shows an issue with the Notary service:

> Performance: Today, 10:10 AM - ongoing
>
> Some users are affected
>
> Users are experiencing a problem with this service. We are working to resolve this issue.

**Comment by andybons** [comment]
Notarization of the installer package has completed successfully. We are testing locally to ensure no scary popups are shown.

**Comment by andybons** [comment]
While we believe we have the installer package issue resolved, we have to deal with the `.tar.gz` bundles, which require notarization of each individual binary. Working on this.

**Comment by andybons** [comment]
As an update, this is what we are going to do:

+ Sign all Mach-O binaries within `$GOROOT/bin/` and `$GOROOT/pkg/tool/darwin_amd64`  (without the Hardened Runtime enabled as it’s not required just yet)
+ Construct the `.pkg` installer with the signed binaries included
+ Sign the `.pkg` installer
+ Send the `.pkg` installer off to the notarization service and staple the resulting ticket to the installer

Since the notarization process is recursive (and the bits being distributed in the `.tar.gz` file are the same that will be in the installer), this should solve the warnings for users that choose to install using either the `.pkg` file or the `.tar.gz` file.

I will update this thread once this work is complete.

**Comment by andybons** [comment]
The initial work mentioned in https://github.com/golang/go/issues/34986#issuecomment-546067381 is now complete. We will likely issue a point release this week with the changes after some further testing.

**Comment by virskor** [comment]
same issue 

```
csrutil status
System Integrity Protection status: disabled.

go version
go version go1.13.3 darwin/amd64

osversion
10.15.1 (19B88)
```

**Comment by andybons** [comment]
New point releases have been released (golang.org/dl/#go1.13.4) that have signed and notarized binaries/installers. From testing locally I did not see any issues but feel free to comment if there are further problems.

Will continue the rest of the work (enabling the Hardened Runtime, for example) on #31918.

**Comment by andybons** [comment]
Oh, and Happy Halloween! 👻

**Comment by andybons** [comment]
Re-opening as #31918 should be tracking whether binaries created by the Go toolchain remain acceptable to Apple’s notarization service, which is a different issue than the toolchain itself being acceptable.

**Comment by andybons** [comment]
Below is derived from the log returned by Apple’s notarization service. I expect any warning to turn into an error come January 2020:

Warnings we intend to fix by enabling the Hardened Runtime on the listed binaries:
```
[warning]: (go/bin/go): The executable does not have the hardened runtime enabled.
[warning]: (go/bin/gofmt): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/vet): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/objdump): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/asm): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/trace): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/pprof): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/dist): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/pack): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/addr2line): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/compile): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/link): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/cgo): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/nm): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/fix): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/test2json): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/doc): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/buildid): The executable does not have the hardened runtime enabled.
[warning]: (go/pkg/tool/darwin_amd64/cover): The executable does not have the hardened runtime enabled.
```

Warnings we’ll have to fix another way:
```
[warning]: (go/src/compress/gzip/testdata/issue6550.gz): b'gunzip: data stream error\ngunzip: /tmp/tmpd_lw2hk0/issue6550.gz.unpacked_00/issue6550.gz: uncompress failed\n'
[warning]: (go/src/archive/zip/testdata/go-no-datadesc-sig.zip): b"ditto: foo.txt: No such file or directory\nditto: Couldn't read pkzip signature.\n"
[warning]: (go/src/cmd/internal/buildid/testdata/a.macho): The binary is not signed.
[warning]: (go/src/cmd/internal/buildid/testdata/a.macho): The signature does not include a secure timestamp.
[warning]: (go/src/cmd/internal/buildid/testdata/a.macho): The executable does not have the hardened runtime enabled.
[warning]: (go/src/debug/macho/testdata/gcc-386-darwin-exec): The binary is not signed.
[warning]: (go/src/debug/macho/testdata/gcc-386-darwin-exec): The signature does not include a secure timestamp.
[warning]: (go/src/debug/macho/testdata/gcc-386-darwin-exec): The executable does not have the hardened runtime enabled.
[warning]: (go/src/debug/macho/testdata/clang-386-darwin-exec-with-rpath): The binary is not signed.
[warning]: (go/src/debug/macho/testdata/clang-386-darwin-exec-with-rpath): The signature does not include a secure timestamp.
[warning]: (go/src/debug/macho/testdata/clang-386-darwin-exec-with-rpath): The executable does not have the hardened runtime enabled.
[warning]: (go/src/debug/macho/testdata/clang-amd64-darwin-exec-with-rpath): The binary is not signed.
[warning]: (go/src/debug/macho/testdata/clang-amd64-darwin-exec-with-rpath): The signature does not include a secure timestamp.
[warning]: (go/src/debug/macho/testdata/clang-amd64-darwin-exec-with-rpath): The executable does not have the hardened runtime enabled.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The binary is not signed.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The signature does not include a secure timestamp.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The executable does not have the hardened runtime enabled.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The binary is not signed.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The signature does not include a secure timestamp.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The executable does not have the hardened runtime enabled.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The binary is not signed.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The signature does not include a secure timestamp.
[warning]: (go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec): The executable does not have the hardened runtime enabled.
[warning]: (go/src/debug/macho/testdata/gcc-amd64-darwin-exec): The binary is not signed.
[warning]: (go/src/debug/macho/testdata/gcc-amd64-darwin-exec): The signature does not include a secure timestamp.
[warning]: (go/src/debug/macho/testdata/gcc-amd64-darwin-exec): The executable does not have the hardened runtime enabled.
```

De-duping from the above, the following binaries need to be dealt with (through obfuscation or some other means):
```
go/src/debug/macho/testdata/clang-amd64-darwin-exec-with-rpath
go/src/archive/zip/testdata/go-no-datadesc-sig.zip
go/src/debug/macho/testdata/fat-gcc-386-amd64-darwin-exec
go/src/debug/macho/testdata/gcc-amd64-darwin-exec
go/src/compress/gzip/testdata/issue6550.gz
go/src/debug/macho/testdata/clang-386-darwin-exec-with-rpath
go/src/cmd/internal/buildid/testdata/a.macho
go/src/debug/macho/testdata/gcc-386-darwin-exec
```

**Comment by aclements** [comment]
Is it saying the binaries need any particular hardened runtime entitlements, or just that all binaries need to be signed as having a hardened runtime?

**Comment by andybons** [comment]
@aclements just that the binaries must have the hardened runtime enabled. It doesn’t (and likely won‘t/can’t?) say what entitlements a binary needs.

**Comment by andybons** [comment]
@gopherbot please open a backport for 1.12 and 1.13 since otherwise Apple will reject future point releases of those versions

**Comment by nightlyone** [comment]
I wonder why tests and test data need to be shipped in releases. People interested in that can always retrieve it from the source repository. Has that been considered?

**Comment by bradfitz** [comment]
@nightlyone, it has, several times. The main reason we continue to include them is that when a user reports a problem, it's nice to be able to ask them "What does `go test syscall` or `go test -v os` say?". That's a question they can easily answer. If we make them do a new download, build, and parallel install, they're very unlikely to do so. Admittedly this debugging procedure doesn't happen too often, though.

**Comment by bcmills** [comment]
It also seems desirable for `go test all` to have the same behavior regardless of how the Go toolchain was installed, particularly in module mode (where `all` is actually a reasonable thing to want to test).

**Comment by andybons** [comment]
I've run an altered `all.bash` that signs (with the [hardened runtime](https://developer.apple.com/documentation/security/hardened_runtime_entitlements) enabled) the toolchain binaries and all tests passed.

We plan to enable the hardened runtime for the 1.14 beta (scheduled for early December) and see if anything breaks. If everything looks good, we'll enable it for all releases moving forward.

**Comment by toothrot** [comment]
Just to reiterate what Andy said, the hardened runtime will be enabled for the go1.14 beta release.

**Comment by cagedmantis** [comment]
I've confirmed with @bcmills that https://github.com/golang/go/issues/36572 should not be considered a blocker for go1.14rc1. I'm going to recommend that we close this issue. @andybons @dmitshur @toothrot

**Comment by dmitshur** [comment]
Sounds good. With the hardened runtime enabled as of Go 1.14 beta 1, I don't believe there's anything left to do here other than to watch for any issues if they're reported, but new issues can be opened for that.

I'll let one of @andybons or @toothrot confirm and close this issue since they're currently assigned to it.

**Comment by andybons** [comment]
Yes this looks to be fixed. If we discover any issues during the notarization stage of a release we'll open a new issue.
